test/cpp/src/701-dram-scheduler.cc

```diff
@@ -0,0 +1,116 @@
+#include <catch.hpp>
+#include "mocks.hpp"
+#include "defaults.hpp"
+#include "dram_controller.h"    
+#include "champsim_constants.h"
+#include <algorithm>
+#include <cfenv>
+#include <cmath>
+
+bool dram_test(MEMORY_CONTROLLER* uut, std::vector<champsim::channel::request_type>* packet_stream, std::vector<uint64_t>* expected_order, std::vector<uint64_t>* arriv_time)
+{
+    uint64_t vector_index = 0;
+    uut->current_cycle = 0;
+
+    std::vector<DRAM_CHANNEL::queue_type::iterator> request_locations;
+    //load requests into controller
+    for(auto it = packet_stream->begin(); it != packet_stream->end(); it++)
+    {
+        auto jt = uut->channels[0].RQ.begin() + vector_index;
+        *jt = DRAM_CHANNEL::request_type{*it};
+        jt->value().forward_checked = false;
+        jt->value().event_cycle = (*arriv_time)[vector_index];
+        request_locations.push_back(jt);
+        vector_index++;
+    }
+
+    //carry out operates, make sure they complete in the expected order
+    uint64_t completed_reqs = 0;
+    std::vector<bool> last_scheduled(packet_stream->size(),false);
+    while(true)
+    {
+        //decide if we are to enqueue a new packet or not
+        uut->operate();
+        uut->current_cycle++;
+        
+        //look through RQ, make sure the right packet was scheduled
+        uint64_t index = 0;
+        for(auto it = uut->channels[0].RQ.begin(); it != uut->channels[0].RQ.end(); it++)
+        {
+            if(it->has_value())
+            if(!last_scheduled[index] && it->has_value() && it->value().scheduled)
+            {
+                
+                if(it != request_locations[(*expected_order)[completed_reqs]])
+                {
+                
+                return(false);
+                }
+                completed_reqs++;
+            }
+            if(it->has_value())
+            {
+                last_scheduled[index] = it->value().scheduled;
+            }
+
+            index++;
+        }
```

Comment:
I had to spend about an hour parsing this. Would it be simpler to do the following?

```suggestion
        std::vector<bool> next_scheduled{};
        std::transform(std::begin(uut->channels[0].RQ), std::end(uut->channels[0].RQ), std::back_inserter(next_scheduled), [](const auto& entry) { return entry.has_value() && entry.value().scheduled; });
        
        auto chunk_begin = std::begin(next_scheduled);
        auto chunk_end = std::end(next_scheduled);
        while (chunk_begin != chunk_end) {
            std::tie(chunk_begin, std::ignored) = std::mismatch(chunk_begin, chunk_end, std::cbegin(last_scheduled));
            if (chunk_begin != chunk_end && chunk_begin != request_locations[(*expected_order)[completed_reqs++]]) {
                return false;
            }
        }
        last_scheduled = next_scheduled;
```